### PR TITLE
Validate tokens

### PR DESCRIPTION
**Pull request by ishangulhane**

### CODE REVIEW

File: lib/src/main/java/com/ibm/bluemix/appid/android/internal/config/Config.java [hunk 1]
```diff
@@ -40,4 +41,12 @@ public static String getUserProfilesServerUrl (AppID appId) {
 		serverUrl += ATTRIBUTES_ENDPOINT;
 		return serverUrl;
 	}
+
+	public static String getPublicKeysEndpoint (AppID appId) {
+		return Config.getOAuthServerUrl(appId)+PUBLIC_KEYS_ENDPOINT;
```
Comment (rotembr) [inline lib/src/main/java/com/ibm/bluemix/appid/android/internal/config/Config.java]: missing spaces between "+"
also I don't see any use of this API 
If this is something users need to use we need to have example in the app module and in the readMe file.

File: lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java [hunk 2]
```diff
@@ -109,21 +130,17 @@ private void retrieveTokens(HashMap<String, String> formParams, final TokenRespo
...
-
+				String errorDescription = "";
 				try {
 					if (response.getStatus() == 400) {
-						JSONObject responseJSON = new JSONObject(response.getResponseText());
-						if (INVALID_GRANT.equals(responseJSON.getString(ERROR))) {
-							listener.onAuthorizationFailure(new AuthorizationException(responseJSON.getString(ERROR_DESCRIPTION)));
-							return;
-						}
-					} else if (response.getStatus() == 403) {
```
Comment (rotembr) [inline lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java]: this code was added by @sandmman to returns pending user error
https://github.com/ibm-cloud-security/appid-clientsdk-android/commit/d6c9022cae37e9cfb141571f24bbc75bb71b1dda
are we sure we need to remove it?

File: lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java [hunk 3]
```diff
@@ -175,55 +192,159 @@ private String createAuthenticationHeader (String clientId) throws Exception {
...
+			case ACCESS_TOKEN: return TOKENS.ID_TOKEN;
+			case ID_TOKEN: return TOKENS.REFRESH_TOKEN;
+			default: return token;
 		}
+	}
+
+	protected void extractTokens(RetrievedTokens retrievedTokens, TokenResponseListener listener) {
+		TOKENS flowStep = retrievedTokens.getFlowStep();
+		if (flowStep == TOKENS.ACCESS_TOKEN || flowStep == TOKENS.ID_TOKEN){
+			logger.debug("Extracting and verifying "+flowStep.toString()+" token from server response");
```
Comment (rotembr) [inline lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java]: spaces between '+' s

File: lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java [hunk 4]
```diff
@@ -175,55 +192,159 @@ private String createAuthenticationHeader (String clientId) throws Exception {
...
+				String kid = token.getHeader().getString("kid");
+				Key tokenPublicKey = getPublicKeyByKid(kid);
+				retrievedTokens.setFlowStep(getNextToken(flowStep));
+				if (tokenPublicKey == null) {
+					lookUpPublicKey(kid, retrievedTokens, listener, tokenType);
+				} else {
+					saveToken(retrievedTokens, tokenPublicKey, listener, tokenType);
+				}
+			} catch (RuntimeException|AuthorizationException|JSONException e){
+				logger.error("Failed to parse "+flowStep.toString()+", error : ", e);
```
Comment (rotembr) [inline lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java]: spaces

File: lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java [hunk 5]
```diff
@@ -175,55 +192,159 @@ private String createAuthenticationHeader (String clientId) throws Exception {
...
+				listener.onAuthorizationFailure(new AuthorizationException("Failed to parse "+flowStep.toString()+", error : "+e.getMessage()));
```
Comment (rotembr) [inline lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java]: spaces

File: lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java (hunk 2 above)
Comment (sandmman) [reply lib/src/main/java/com/ibm/bluemix/appid/android/internal/tokenmanager/TokenManager.java]: @ishangulhane This should not be removed. We need to propagate the 403 error. Seems like some merge got your branch out of sync.

File: lib/src/main/java/com/ibm/bluemix/appid/android/internal/config/Config.java (hunk 1 above)
Comment (ishangulhane) [reply lib/src/main/java/com/ibm/bluemix/appid/android/internal/config/Config.java]: It is used for getting the public keys in the lookUpPublicKey method in TokenManager.

**Review by gtaban — COMMENTED** [review]
Logic looks good, can you also get another review? thanks.
Also why the new cradle package options?

**Comment by ishangulhane** [comment]
Sometimes app build fails with `Duplicate files copied in APK META-INF/xxx`. Added package options to fix the issue.

**Review by gtaban — APPROVED** [review]

**Review by rotembr — APPROVED** [review]
looks good overall.
Please make sure everything continue to work as usual, and run our automation tests with this Android SDK changes.

**Pull request merged**